<a href="https://colab.research.google.com/github/Tessellate-Imaging/monk_v1/blob/master/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/5)%20Feature%20-%20Switch%20modes%20without%20reloading%20experiment%20-%20train%2C%20eval%2C%20infer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goals

### 1. Understand how continuously and manually switching between training and externally validating help improve training

### 2. Experiment and learning how switching between train and val modes can help choose right hyper-parameters


### 2. Steps
    - You will use mxnet gluon backend for this example
    - You will first train a classifier using default params
    
    - You will switch mode from train to val for the first time here
    - You will then validate to check accuracy
    
    - You will switch mode from val to train here
    - You will reduce the learning rate (Need notfocus on how it is done for now)
    - You will retrain again using this new lr
    
    - You will switch mode from train to val for the second time here
    - You will then validate to check accuracy
    
    - You will again switch mode from val to train here
    - You will further change the learning rate (Need notfocus on how it is done for now)
    - You will retrain again using this newest lr
    
    - You will switch mode from train to val for the final time here
    - You will then validate to check accuracy

# Table of Contents


## [Install](#0)


## [Train a classifier using default settings](#1)


## [Switch mode from train to eval and validate](#2)


## [Switch back mode, reduce lr, retrain](#3)


## [Switch mode from train to eval and re-validate](#4)


## [Switch back mode, change lr further, retrain](#5)


## [Switch mode from train to eval and re-validate](#6)

<a id='0'></a>
# Install Monk

## Using pip (Recommended)

  - colab (gpu) 
      - All bakcends: `pip install -U monk-colab`
      

  - kaggle (gpu) 
      - All backends: `pip install -U monk-kaggle`
      

  - cuda 10.2	
      - All backends: `pip install -U monk-cuda102`
      - Gluon bakcned: `pip install -U monk-gluon-cuda102`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda102`
      - Keras backend: `pip install -U monk-keras-cuda102`
      

  - cuda 10.1	
      - All backend: `pip install -U monk-cuda101`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda101`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda101`
	  - Keras backend: `pip install -U monk-keras-cuda101`
      

  - cuda 10.0	
      - All backend: `pip install -U monk-cuda100`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda100`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda100`
	  - Keras backend: `pip install -U monk-keras-cuda100`
      

  - cuda 9.2	
      - All backend: `pip install -U monk-cuda92`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda92`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda92`
	  - Keras backend: `pip install -U monk-keras-cuda92`
      

  - cuda 9.0	
      - All backend: `pip install -U monk-cuda90`
	  - Gluon bakcned: `pip install -U monk-gluon-cuda90`
	  - Pytorch backend: `pip install -U monk-pytorch-cuda90`
	  - Keras backend: `pip install -U monk-keras-cuda90`
      

  - cpu 		
      - All backend: `pip install -U monk-cpu`
	  - Gluon bakcned: `pip install -U monk-gluon-cpu`
	  - Pytorch backend: `pip install -U monk-pytorch-cpu`
	  - Keras backend: `pip install -U monk-keras-cpu`

## Install Monk Manually (Not recommended)
 
### Step 1: Clone the library
 - git clone https://github.com/Tessellate-Imaging/monk_v1.git
 
 
 
 
### Step 2: Install requirements 
 - Linux
     - Cuda 9.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu90.txt`
     - Cuda 9.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu92.txt`
     - Cuda 10.0
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu100.txt`
     - Cuda 10.1
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu101.txt`
     - Cuda 10.2
         - `cd monk_v1/installation/Linux && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Linux && pip install -r requirements_cpu.txt`
 
 
  - Windows
     - Cuda 9.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu90.txt`
     - Cuda 9.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu92.txt`
     - Cuda 10.0 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu100.txt`
     - Cuda 10.1 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu101.txt`
     - Cuda 10.2 (Experimental support)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cu102.txt`
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Windows && pip install -r requirements_cpu.txt`
         
         
  - Mac
     - CPU (Non gpu system)
         - `cd monk_v1/installation/Mac && pip install -r requirements_cpu.txt`
         
         
  - Misc
      - Colab (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_colab.txt`
      - Kaggle (GPU)
          - `cd monk_v1/installation/Misc && pip install -r requirements_kaggle.txt`
 
 
 
### Step 3: Add to system path (Required for every terminal or kernel run)
 - `import sys`
 - `sys.path.append("monk_v1/");`

## Dataset - Malarial cell images
    - Credits: https://www.kaggle.com/iarunava/cell-images-for-detecting-malaria

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1mMEtGIK8UZNCrErXRJR-kutNTaN1zxjC" -O malaria_cell.zip && rm -rf /tmp/cookies.txt

In [ ]:
! unzip -qq malaria_cell.zip

In [ ]:
! wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1WHpd7M-E_EiXmdjOr48BfvlUtMRPV6PM" -O malaria_cell_val.zip && rm -rf /tmp/cookies.txt

In [8]:
! unzip -qq malaria_cell_val.zip

# Imports 

    - Using single mxnet-gluoncv backend for this tutorial

In [1]:
#Using gluon backend 

# When installed using pip
from monk.gluon_prototype import prototype


# When installed manually (Uncomment the following)
#import os
#import sys
#sys.path.append("monk_v1/");
#sys.path.append("monk_v1/monk/");
#from monk.gluon_prototype import prototype

<a id='1'></a>
# Train a classifier using default settings

### Creating and managing experiments
        - Provide project name
        - Provide experiment name

In [2]:
gtf = prototype(verbose=1);
gtf.Prototype("Malaria-Cell", "exp-switch-modes");

Mxnet Version: 1.5.1

Experiment Details
    Project: Malaria-Cell
    Experiment: exp-switch-modes
    Dir: /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/



### This creates files and directories as per the following structure
    
    
    workspace
        |
        |--------Malaria-Cell
                        |
                        |
                        |-----exp-switch-modes
                                    |
                                    |-----experiment-state.json
                                    |
                                    |-----output
                                            |
                                            |------logs (All training logs and graphs saved here)
                                            |
                                            |------models (all trained models saved here)

### Load Dataset

In [3]:
gtf.Default(dataset_path="malaria_cell", 
            model_name="resnet18_v1", 
            num_epochs=5);

#Read the summary generated once you run this cell. 

Dataset Details
    Train path:     malaria_cell
    Val path:       None
    CSV train path: None
    CSV val path:   None
    Label Type:     single

Dataset Params
    Input Size:   224
    Batch Size:   4
    Data Shuffle: True
    Processors:   8
    Train-val split:   0.7

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Params
    Model name:           resnet18_v1
    Use Gpu:              True
    Use pretrained:       True
    Freeze base network:  True

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  41
        Nu

/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: clipnorm and clipvalue are active only for keras in current version of Monk
  warnings.warn(msg)
/home/ubuntu/.virtualenvs/monk_pip_test/lib/python3.6/site-packages/monk/system/imports.py:160: UserWarning: ArgumentWarning: momentum_dampening_rate is active only for pytorch in current version of Monk
  warnings.warn(msg)


### From summary current Learning rate: 0.01

In [4]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------





    curr_lr - 0.01
    [Epoch 1] Train-acc: 0.725, Train-loss: 0.662 | Val-acc: 0.804959, Val-loss: 0.470, | time: 4.0 sec

    Epoch 2/5
    ----------





    curr_lr - 0.0098
    [Epoch 2] Train-acc: 0.753, Train-loss: 0.630 | Val-acc: 0.813223, Val-loss: 0.459, | time: 4.0 sec

    Epoch 3/5
    ----------





    curr_lr - 0.009604
    [Epoch 3] Train-acc: 0.790, Train-loss: 0.536 | Val-acc: 0.808264, Val-loss: 0.502, | time: 3.9 sec

    Epoch 4/5
    ----------





    curr_lr - 0.009411919999999999
    [Epoch 4] Train-acc: 0.784, Train-loss: 0.543 | Val-acc: 0.657851, Val-loss: 1.071, | time: 3.9 sec

    Epoch 5/5
    ----------





    curr_lr - 0.009223681599999999
    [Epoch 5] Train-acc: 0.803, Train-loss: 0.499 | Val-acc: 0.849587, Val-loss: 0.371, | time: 4.0 sec

    Training completed in: 0m 18s
    Best val Acc:          0.849587

Training End

Training Outputs
    Model Dir:   /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training curve:                  train_loss_history.npy
  

<Figure size 432x288 with 0 Axes>

<a id='2'></a>
# Switch mode from train to eval and validate

In [5]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [9]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   8

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [10]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 59.76331360946746 %
            1. Uninfected - 61.111111111111114 %
        total images:            1011
        num correct predictions: 611
        Average accuracy (%):    60.43521266073195



### Accuracy now is - 60.43% when learning rate is 0.01
(Can change when you run the exp)

<a id='3'></a>
# Switch back mode, reduce lr, retrain

In [11]:
gtf.Switch_Mode(train=True)

## Reduce learning rate from 0.01 to 0.001

In [12]:
# This part of code will be taken up again in upcoming sections
gtf.update_learning_rate(0.001);
gtf.Reload();

Update: Learning Rate - 0.001

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  41
        Num of actual trainable layers:       1



In [13]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------




    curr_lr - 0.009039207967999998
    [Epoch 1] Train-acc: 0.686, Train-loss: 0.601 | Val-acc: 0.786777, Val-loss: 0.445, | time: 4.3 sec

    Epoch 2/5
    ----------





    curr_lr - 0.008858423808639998
    [Epoch 2] Train-acc: 0.775, Train-loss: 0.473 | Val-acc: 0.738843, Val-loss: 0.511, | time: 3.9 sec

    Epoch 3/5
    ----------





    curr_lr - 0.008681255332467197
    [Epoch 3] Train-acc: 0.806, Train-loss: 0.418 | Val-acc: 0.829752, Val-loss: 0.378, | time: 4.4 sec

    Epoch 4/5
    ----------




    curr_lr - 0.008507630225817853
    [Epoch 4] Train-acc: 0.792, Train-loss: 0.444 | Val-acc: 0.798347, Val-loss: 0.401, | time: 4.0 sec

    Epoch 5/5
    ----------




    curr_lr - 0.008337477621301496
    [Epoch 5] Train-acc: 0.810, Train-loss: 0.422 | Val-acc: 0.809917, Val-loss: 0.433, | time: 3.9 sec

    Training completed in: 0m 38s
    Best val Acc:          0.829752

Training End

Training Outputs
    Model Dir:   /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training curve:                  train_loss_history.npy
   

<Figure size 432x288 with 0 Axes>

<a id='4'></a>
# Switch mode from train to eval and re-validate

In [14]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [15]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   8

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [16]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 83.23471400394477 %
            1. Uninfected - 26.785714285714285 %
        total images:            1011
        num correct predictions: 557
        Average accuracy (%):    55.093966369930754



### Accuracy now is - 55.09% when learning rate is 0.001 
(Can change when you run the exp)
    - Thus reducing learning rate didn't help our case

<a id='5'></a>
# Switch back mode, change lr, retrain

In [17]:
gtf.Switch_Mode(train=True)

## Update the learning rate again

In [18]:
# This part of code will be taken up again in upcoming sections
gtf.update_learning_rate(0.1);
gtf.Reload();

Update: Learning Rate - 0.1

Pre-Composed Train Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Pre-Composed Val Transforms
[{'RandomHorizontalFlip': {'p': 0.8}}, {'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num train images: 1411
    Num val images:   605
    Num classes:      2

Model Details
    Loading pretrained model
    Model Loaded on device
        Model name:                           resnet18_v1
        Num of potentially trainable layers:  41
        Num of actual trainable layers:       1



In [19]:
#Start Training
gtf.Train();

#Read the training summary generated once you run the cell and training is completed

Training Start
    Epoch 1/5
    ----------





    curr_lr - 0.008170728068875466
    [Epoch 1] Train-acc: 0.549, Train-loss: 0.789 | Val-acc: 0.604959, Val-loss: 0.670, | time: 4.1 sec

    Epoch 2/5
    ----------




    curr_lr - 0.008007313507497957
    [Epoch 2] Train-acc: 0.678, Train-loss: 0.603 | Val-acc: 0.715702, Val-loss: 0.554, | time: 4.1 sec

    Epoch 3/5
    ----------




    curr_lr - 0.007847167237347998
    [Epoch 3] Train-acc: 0.719, Train-loss: 0.545 | Val-acc: 0.720661, Val-loss: 0.534, | time: 4.0 sec

    Epoch 4/5
    ----------





    curr_lr - 0.007690223892601038
    [Epoch 4] Train-acc: 0.756, Train-loss: 0.485 | Val-acc: 0.755372, Val-loss: 0.501, | time: 4.1 sec

    Epoch 5/5
    ----------





    curr_lr - 0.007536419414749017
    [Epoch 5] Train-acc: 0.773, Train-loss: 0.477 | Val-acc: 0.778512, Val-loss: 0.479, | time: 4.1 sec

    Training completed in: 0m 58s
    Best val Acc:          0.778512

Training End

Training Outputs
    Model Dir:   /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/models/
    Log Dir:     /home/ubuntu/Desktop/monk_pip_test/monk_v1/study_roadmaps/1_getting_started_roadmap/2_elemental_features_of_monk/workspace/Malaria-Cell/exp-switch-modes/output/logs/
    Final model: final
    Best model:  best_model
    Log 1 - Validation accuracy history log: val_acc_history.npy
    Log 2 - Validation loss history log:     val_loss_history.npy
    Log 3 - Training accuracy history log:   train_acc_history.npy
    Log 4 - Training loss history log:       train_loss_history.npy
    Log 5 - Training curve:                  train_loss_history.npy
  

<Figure size 432x288 with 0 Axes>

<a id='6'></a>
# Switch mode from train to eval and re-validate

In [20]:
gtf.Switch_Mode(eval_infer=True)

### Load the validation dataset

In [21]:
gtf.Dataset_Params(dataset_path="malaria_cell_val");
gtf.Dataset();

Dataset Details
    Test path:      malaria_cell_val
    CSV test path:  None

Dataset Params
    Input Size:  224
    Processors:   8

Pre-Composed Test Transforms
[{'Normalize': {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}}]

Dataset Numbers
    Num test images: 1011
    Num classes:      2



### Run validation

In [22]:
accuracy, class_based_accuracy = gtf.Evaluate();

Testing



    Result
        class based accuracies
            0. Parasitized - 49.5069033530572 %
            1. Uninfected - 48.80952380952381 %
        total images:            1011
        num correct predictions: 497
        Average accuracy (%):    49.15924826904055



### Accuracy now is - 49.15% when learning rate is 0.1, even  lower
(Can change when you run the exp)
    - Thus increasing learning rate didn't help our case

### LR 0.01 worked best for us
    - That's how manual hyper-parameter tuning can be done using switch modes

# Goals Completed

### 1. Understand how continuously and manually switching between training and externally validating help improve training

### 2. Experiment and learning how switching between train and val modes can help choose right hyper-parameters


### 3. Steps completed
    - use mxnet gluon backend for this example
    - train a classifier using default params
    
    - switch mode from train to val for the first time here
    - validate to check accuracy
    
    - switch mode from val to train here
    - reduce the learning rate (Need notfocus on how it is done for now)
    - retrain again using this new lr
    
    - switch mode from train to val for the second time here
    - validate to check accuracy
    
    - switch mode from val to train here
    - change the learning rate (Need notfocus on how it is done for now)
    - retrain again using this newest lr
    
    - switch mode from train to val for the final time here
    - validate to check accuracy